src/tree_builder/mod.rs

```diff
@@ -435,3 +450,157 @@ impl<Handle, Sink> TokenSink
+
+    pub struct LineCountingDOM {
+        pub line_vec: Vec<(QualName, u64)>,
+        pub rcdom: RcDom,
+    }
+
+    impl TreeSink for LineCountingDOM {
+        type Output = Self;
+
+        fn finish(self) -> Self { self }
+
+        type Handle = Handle;
+
+        fn parse_error(&mut self, msg: Cow<'static, str>) {
+            self.rcdom.parse_error(msg);
+        }
+
+        fn get_document(&mut self) -> Handle {
+            self.rcdom.get_document()
+        }
+
+        fn get_template_contents(&mut self, target: Handle) -> Handle {
+            self.rcdom.get_template_contents(target)
+        }
+
+        fn set_quirks_mode(&mut self, mode: QuirksMode) {
+            self.rcdom.set_quirks_mode(mode)
+        }
+
+        fn same_node(&self, x: Handle, y: Handle) -> bool {
+            self.rcdom.same_node(x, y)
+        }
+
+        fn elem_name(&self, target: Handle) -> QualName {
+            self.rcdom.elem_name(target)
+        }
+
+        fn create_element(&mut self, name: QualName, attrs: Vec<Attribute>) -> Handle {
+            self.line_vec.push((name.clone(), self.rcdom.current_line));
+            self.rcdom.create_element(name, attrs)
+        }
+
+        fn create_comment(&mut self, text: StrTendril) -> Handle {
+            self.rcdom.create_comment(text)
+        }
+
+        fn append(&mut self, parent: Handle, child: NodeOrText<Handle>) {
+            self.rcdom.append(parent, child)
+        }
+
+        fn append_before_sibling(&mut self,
+                sibling: Handle,
+                child: NodeOrText<Handle>) -> Result<(), NodeOrText<Handle>> {
+            self.rcdom.append_before_sibling(sibling, child)
+        }
+
+        fn append_doctype_to_document(&mut self,
+                                      name: StrTendril,
+                                      public_id: StrTendril,
+                                      system_id: StrTendril) {
+            self.rcdom.append_doctype_to_document(name, public_id, system_id);
+        }
+
+        fn add_attrs_if_missing(&mut self, target: Handle, attrs: Vec<Attribute>) {
+            self.rcdom.add_attrs_if_missing(target, attrs);
+        }
+
+        fn remove_from_parent(&mut self, target: Handle) {
+            self.rcdom.remove_from_parent(target);
+        }
+
+        fn reparent_children(&mut self, node: Handle, new_parent: Handle) {
+            self.rcdom.reparent_children(node, new_parent);
+        }
+
+        fn mark_script_already_started(&mut self, target: Handle) {
+            self.rcdom.mark_script_already_started(target);
+        }
+
+        fn is_mathml_annotation_xml_integration_point(&self, handle: Self::Handle) -> bool {
+            self.rcdom.is_mathml_annotation_xml_integration_point(handle)
+        }
+
+        fn set_current_line(&mut self, line_number: u64) {
+            self.set_current_line(line_number);
+            self.rcdom.set_current_line(line_number);
+        }
+    }
+
+    impl Default for LineCountingDOM {
```

Comment:
We don't need this implementation; we can write out the initializer by hand instead of using LineCountingDOM::default().